java/arcs/core/storage/DirectStore.kt

```diff
@@ -109,23 +109,19 @@ class DirectStore<Data : CrdtData, Op : CrdtOperation, T> /* internal */ constru
     }
 
     /** Closes the store. Once closed, it cannot be re-opened. A new instance must be created. */
-    override fun close() {
-        synchronized(proxyManager) {
+    override suspend fun close() {
+        val needToClose = synchronized(proxyManager) {
```

Comment:
I think it's generally considered a no-no to use synchonized inside of a coroutine. Should use mutex.withLock { ... } instead

Reply:
https://blog.danlew.net/2020/01/28/coroutines-and-java-synchronization-dont-mix/

Reply:
Oh, I just saw you re-entrancy and mutexes aren't re-entrant.  Maybe use a channel + flow?

Reply:
I don't think using synchronized in a coroutine is an issue, except that you may block the thread, defeating the point of using a coroutine. 

Suspending inside of a critical section is a big issue (and the IDE will warn you about it). But that's not happening here. 

A mutex can't be used here because there are non-suspending methods that need to access the protected element.

Reply:
From your comments and @ianull 's, I have thoughts on a second pass here, which I'll do tomorrow.